src/span.h

```diff
@@ -18,12 +26,29 @@ template<typename C>
 class Span
 {
     C* m_data;
-    std::ptrdiff_t m_size;
+    std::size_t m_size;
 
 public:
     constexpr Span() noexcept : m_data(nullptr), m_size(0) {}
-    constexpr Span(C* data, std::ptrdiff_t size) noexcept : m_data(data), m_size(size) {}
-    constexpr Span(C* data, C* end) noexcept : m_data(data), m_size(end - data) {}
+
+    /** Construct a span from a begin pointer and a size.
+     *
+     * This implements a subset of the iterator-based std::span constructor in C++20,
+     * which is hard to implement without std::address_of.
+     */
+    template <typename T, typename std::enable_if<std::is_convertible<T (*)[], C (*)[]>::value, int>::type = 0>
+    constexpr Span(T* begin, std::size_t size) noexcept : m_data(begin), m_size(size) {}
```

Comment:
At least in C++11, because you're telling it what `C` is, perhaps without knowing it.

You could invoke this constructor like this:

```c++
int arr[5] = {1, 2, 3, 4, 5};
Span<const int> sp{&arr[1], 3};
```

Which is actually a conversion, as `&arr[1]` has type `T`=`int`, but `C`=`const int`.

In other contexts, you may not know that `C` is known. For example:

```c++
void Foo(Span<const int> arg);
...
int arr[5] = {1, 2, 3, 4, 5};
Foo({&arr[1], 3});
```

where `C` is again `const int`, but it's implied by the argument type of the called function.